File: app/src/main/java/i/am/shiro/amai/fragment/MainFragment.java
package i.am.shiro.amai.fragment;

import android.os.Bundle;
import android.os.SystemClock;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.snackbar.Snackbar;

import i.am.shiro.amai.R;
import i.am.shiro.amai.widget.OnBackPressListener;

import static androidx.core.view.ViewCompat.requireViewById;

public final class MainFragment extends Fragment {

    private long lastBackPressTime;

    public MainFragment() {
        super(R.layout.fragment_main);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        view.requestFocus();
        view.setOnKeyListener(new OnBackPressListener(this::onBackPress));

        BottomNavigationView navigation = requireViewById(view, R.id.navigation);
        navigation.setSelectedItemId(R.id.navigation_nhentai);
        navigation.setOnNavigationItemSelectedListener(this::onNavigationItemSelected);

        FragmentManager childFragmentManager = getChildFragmentManager();
        if (childFragmentManager.findFragmentById(R.id.fragmentContainer) == null) {
            NhentaiFragment initialFragment = new NhentaiFragment();
            String fragmentTag = NhentaiFragment.class.getName();
            childFragmentManager.beginTransaction()
                    .add(R.id.fragmentContainer, initialFragment, fragmentTag)
                    .setPrimaryNavigationFragment(initialFragment)
                    .commit();
        }
    }

    private void onBackPress() {
        long now = SystemClock.elapsedRealtime();
        if (now > lastBackPressTime + 1000) {
            lastBackPressTime = now;
            Snackbar.make(requireView(), R.string.confirm_exit, Snackbar.LENGTH_SHORT)
                .setAnchorView(R.id.navigation)
                .show();
        } else {
            requireActivity().finish();
        }
    }

    private boolean onNavigationItemSelected(@NonNull MenuItem item) {
        FragmentManager fragmentManager = getChildFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

        // detach previous fragment if any
        Fragment detachingFragment = fragmentManager.getPrimaryNavigationFragment();
        if (detachingFragment != null) {
            fragmentTransaction.detach(detachingFragment);
        }

        int itemId = item.getItemId();
        String fragmentTag = getFragmentTag(itemId);
        Fragment attachingFragment = fragmentManager.findFragmentByTag(fragmentTag);
        if (attachingFragment == null) {
            // add fragment if it has not been added
            attachingFragment = getFragment(itemId);
            fragmentTransaction.add(R.id.fragmentContainer, attachingFragment, fragmentTag);
        } else {
            // attach fragment if it has already been added
            fragmentTransaction.attach(attachingFragment);
        }
        fragmentTransaction.setPrimaryNavigationFragment(attachingFragment);

        fragmentTransaction.setReorderingAllowed(true);
        fragmentTransaction.commit();
        return true;
    }

    private String getFragmentTag(@IdRes int itemId) {
        switch (itemId) {
            case R.id.navigation_saved:
                return SavedFragment.class.getName();
            case R.id.navigation_nhentai:
                return NhentaiFragment.class.getName();
            case R.id.navigation_downloads:
                return DownloadsFragment.class.getName();
            default:
                throw new IllegalArgumentException("No corresponding fragment for given itemId");
        }
    }

    private Fragment getFragment(@IdRes int itemId) {
        switch (itemId) {
            case R.id.navigation_saved:
                return new SavedFragment();
            case R.id.navigation_nhentai:
                return new NhentaiFragment();
            case R.id.navigation_downloads:
                return new DownloadsFragment();
            default:
                throw new IllegalArgumentException("No corresponding fragment for given itemId");
        }
    }
}

File: app/src/main/java/i/am/shiro/amai/fragment/MainFragment.kt
package i.am.shiro.amai.fragment

import android.os.Bundle
import android.os.SystemClock
import android.view.View
import androidx.activity.addCallback
import androidx.fragment.app.Fragment
import androidx.fragment.app.commit
import androidx.fragment.app.commitNow
import com.google.android.material.snackbar.Snackbar
import i.am.shiro.amai.R
import i.am.shiro.amai.RESULT_TAG
import i.am.shiro.amai.databinding.FragmentMainBinding

class MainFragment : Fragment(R.layout.fragment_main) {

    private var lastBackPressTime = 0L

    private lateinit var savedFragment: SavedFragment

    private lateinit var nhentaiFragment: Nhentai2Fragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val savedTag = "saved"
        val nhentaiTag = "nhentai"

        if (savedInstanceState == null) {
            savedFragment = SavedFragment()
            nhentaiFragment = Nhentai2Fragment()

            childFragmentManager.commitNow {
                add(R.id.fragmentContainer, savedFragment, savedTag)
                add(R.id.fragmentContainer, nhentaiFragment, nhentaiTag)
                detach(savedFragment)
            }
        } else {
            savedFragment = childFragmentManager.findFragmentByTag(savedTag) as SavedFragment
            nhentaiFragment = childFragmentManager.findFragmentByTag(nhentaiTag) as Nhentai2Fragment
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val b = FragmentMainBinding.bind(view)

        requireActivity().onBackPressedDispatcher.addCallback(viewLifecycleOwner) {
            val now = SystemClock.elapsedRealtime()
            if (now > lastBackPressTime + 1000) {
                lastBackPressTime = now
                // TODO replace this with actual view in layout
                Snackbar.make(view, R.string.confirm_exit, Snackbar.LENGTH_SHORT)
                    .setAnchorView(b.navigation)
                    .show()
            } else {
                requireActivity().finish()
            }
        }

        b.navigation.selectedItemId = when {
            !savedFragment.isDetached -> R.id.navigation_nhentai
            !nhentaiFragment.isDetached -> R.id.navigation_nhentai
            else -> error(childFragmentManager.fragments)
        }
        b.navigation.setOnItemSelectedListener {
            onNavigate(it.itemId)
            true
        }

        parentFragmentManager.setFragmentResultListener(RESULT_TAG, viewLifecycleOwner) { _, result ->
            b.navigation.selectedItemId = R.id.navigation_nhentai
            childFragmentManager.setFragmentResult(RESULT_TAG, result)
        }
    }

    private fun onNavigate(itemId: Int) {
        val targetFragment = when (itemId) {
            R.id.navigation_saved -> savedFragment
            R.id.navigation_nhentai -> nhentaiFragment
            else -> throw IllegalArgumentException()
        }

        childFragmentManager.commit {
            for (fragment in arrayOf(savedFragment, nhentaiFragment)) {
                if (fragment == targetFragment) attach(fragment) else detach(fragment)
            }
            setReorderingAllowed(true)
        }
    }
}
